api/TransportTypes.ts

```diff
@@ -0,0 +1,116 @@
+import { ZodType, z } from "zod";
+import { Handler, Middleware } from "../lib/Middleware";
+import { JSONSerializableValue, NonEmptyPartial } from "../lib/Types/HelperTypes";
+import { StrictExtends } from "../lib/Types/StrictExtendsType";
+import { URLEndpoint, URLParameters } from "../lib/URL";
+
+export type HTTPMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE"
+
+export type StatusCodeMap = {
+  status200: 200
+  status201: 201
+  status204: 204
+  status400: 400
+  status401: 401
+  status403: 403
+  status404: 404
+  status429: 429
+  status500: 500
+}
+
+export type StatusCodes = StatusCodeMap[keyof StatusCodeMap]
+
+type APIResponseSchemas = NonEmptyPartial<{
+  [key in keyof StatusCodeMap]: key extends "status204"
+    ? "no-content"
+    : ZodType
+}>
+
+export type TiFAPIResponse<Schemas extends APIResponseSchemas> = {
+  [key in keyof Schemas]: key extends "status204"
+    ? {
+      status: 204,
+      data?: undefined
+    }
+    : Schemas[key] extends ZodType ? {
+        status: StatusCodeMap[key & keyof StatusCodeMap]
+        data: z.infer<Schemas[key]>
+      }
+    : never
+}[keyof Schemas]
+
+type TiFAPIInput = {
+  body?: { [key: string]: JSONSerializableValue };
+  query?: URLParameters;
+  params?: URLParameters;
+};
+
+export type TiFAPIInputContext<T> = {
+  endpointName: string;
+  endpointSchema: GenericEndpointSchema;
+} & T & TiFAPIInput;
+
+export type GenericEndpointSchema = {
+  input: InputSchema;
+  outputs: APIResponseSchemas;
+  constraints?: (input: any, output: any) => boolean;
+  httpRequest: HttpRequest<HTTPMethod>;
+}
+
+/**
+ * Generic API endpoint middleware
+ */
+export type APIMiddleware<T = {}> = Middleware<TiFAPIInputContext<T>, TiFAPIResponse<any>>
+
+/**
+ * Generic API endpoint function
+ */
+export type APIHandler<T = {}> = Handler<TiFAPIInputContext<T>, TiFAPIResponse<any>>;
+
+/**
+ * Type asserts an endpoint schema.
+ */
+export const assertEndpointSchemaType = <
```

Comment:
Nit: Would probably change the name of this, when creating the schema it sounds like your writing a test.